backend/src/Edna/DB/Integration.hs

```diff
@@ -30,21 +29,24 @@ import Database.Beam.Query
   (SqlDelete, SqlInsert, SqlSelect, SqlUpdate, runDelete, runInsert, runSelectReturningList,
   runSelectReturningOne, runUpdate)
 import Database.Beam.Schema (Beamable)
-import Database.PostgreSQL.Simple.Transaction (withTransactionSerializable)
-import RIO (withRunInIO)
 
-import Edna.DB.Connection (ConnPool(..))
+import Edna.DB.Connection (PostgresConn(..), postgresConnSingle)
 import Edna.Logging (logUnconditionally)
-import Edna.Setup (Edna, edConnectionPool, edDebugDB)
+import Edna.Setup (Edna, EdnaContext(_edDBConnection), edDBConnection, edDebugDB)
 
 withConnection :: (Connection -> Edna a) -> Edna a
 withConnection action = do
-  ConnPool pool <- view edConnectionPool
-  withRunInIO $ \unlift -> withResource pool (unlift . action)
+  c <- view edDBConnection
+  _withConnection c action
 
 transact :: Edna a -> Edna a
-transact action = withConnection $
-  \conn -> withRunInIO $ \unlift -> withTransactionSerializable conn (unlift action)
+transact action = do
+  c <- view edDBConnection
+  _withConnection c $ \conn ->
+    _withTransaction c conn $ do
+      local (\ctx -> ctx { _edDBConnection = postgresConnSingle conn }) $ do
```

Comment:
Might be simpler to write `let c' = `postgresConnSingle conn`, then `ctx { _edDBConnection = c' }` here and just use `c'` below (without `view`). But that's up to you.